### ISSUE TITLE
support cmap parameter for line plots

### ISSUE THREAD

**Issue by knaaptime** [holoviz/hvplot#652]
I have a dataframe `data` with three dimensions (x,y,z) and I'd like to create a line plot where the color of the line is set by a colormap rather than randomly-chosen colors.

currently I can do
```python
data.hvplot(x='x', y='y', by='z', cmap='Reds')
```

which gives nearly the appropriate plot, except that the `cmap` parameter is ignored. This causes problems when `z` is a continuous variable rather than a categorical variable because in that case, a sequential colormap is necessary to shade each line appropriately (instead of the default categorical colormap). I'd like for the `cmap` parameter to be recognized in line graphs as it is with other types, e.g. `scatter`

**Comment by maximlt** [comment]
This seems to me like a very nice feature to have in hvPlot! This is actually currently possible, using `.opts` and some HoloViews-fu, applying the options to the elements contained by the resulting NdOverlay:

```python
import holoviews as hv
import hvplot.pandas
import numpy as np
import pandas as pd

df = pd.DataFrame({
  'x': list(np.arange(10)) * 6,
  'y': np.concatenate([np.arange(10) + x for x in range(6)]),
  'z': [l for l in range(0, 60, 10) for _ in range(10)],
})

df.hvplot.line('x', 'y', by='z').opts(hv.opts.Curve(color=hv.Palette('Reds')))
```

<img width="700" alt="image" src="https://user-images.githubusercontent.com/35924738/198365268-e62b1942-5644-4602-bbc3-8bfe7b3839e4.png">

Any thought on this @jlstevens @jbednar  ? This is right now in the `Wishlist` milestone so bound to best ignored indefinitely. I feel this should be supported by hvPlot but one could make the case that it's just a matter of better documenting `.opts` to hvPlot users.

By the way the fact that passing `cmap` is swallowed without any effect or error/warning is a bug, I'm going to open an issue for that.

**Comment by LecrisUT** [comment]
Does anyone have a quick update on the workaround? With the latest version it complains that `hvplot.opts` is not defined

**Comment by maximlt** [comment]
I hit this issue again when working on documenting hvPlot's compatibility with Pandas API.

```python
import pandas as pd
import numpy as np

ts = pd.Series(np.random.randn(1000), index=pd.date_range('1/1/2000', periods=1000))
ts = ts.cumsum()
df = pd.DataFrame(np.random.randn(1000, 10), index=ts.index)
df = df.cumsum()
df.plot(colormap="cubehelix", backend='matplotlib')
```

![Image](https://github.com/user-attachments/assets/0ceaba80-8fc8-4394-9e81-cba02987ce4a)

The same code using `cmap`/`colormap` with hvPlot simply ignores the passed colormap:

![Image](https://github.com/user-attachments/assets/491d0000-e5d5-415a-90ca-d114112e5775)

This can be worked around by setting `color` with an `hv.Palette` object:

```python
import hvplot.pandas
import holoviews as hv
import matplotlib
import pandas as pd
import numpy as np

# cubehelix palette only added when the matplotlib extension is loaded
hv.Palette.colormaps['cubehelix'] = matplotlib.colormaps['cubehelix']

ts = pd.Series(np.random.randn(1000), index=pd.date_range('1/1/2000', periods=1000))
ts = ts.cumsum()
df = pd.DataFrame(np.random.randn(1000, 10), index=ts.index)
df = df.cumsum()
df.hvplot(color=hv.Palette("cubehelix"))
```

![Image](https://github.com/user-attachments/assets/93223cfd-2eb0-4e97-9213-b2d10ceca02b)

---

The code I suggested a couple years ago can be simplified a bit:

```python
import holoviews as hv
import hvplot.pandas
import numpy as np
import pandas as pd

df = pd.DataFrame({
  'x': list(np.arange(10)) * 6,
  'y': np.concatenate([np.arange(10) + x for x in range(6)]),
  'z': [l for l in range(0, 60, 10) for _ in range(10)],
})

df.hvplot.line('x', 'y', by='z', color=hv.Palette('Reds'))
```

**Comment by jbednar** [comment]
@maximlt , would you propose that `cmap` or `colormap` does the same thing as `color` here? If so that sounds good, and seems like a PR that would improve people's lives!

**Comment by maximlt** [comment]
Yes, and without having to use the `hv.Palette` object.

@Azaya89 is this something you recorded in your notebook that documents all the ways users can color things in hvPlot?

**Comment by Azaya89** [comment]
I think there are multiple issues here...

In the dummy dataframe below, the column being used to color the line plot is a numerical column ('z'):
```python
import hvplot.pandas
import numpy as np
import pandas as pd

df = pd.DataFrame({
  'x': list(np.arange(10)) * 6,
  'y': np.concatenate([np.arange(10) + x for x in range(6)]),
  'z': [l for l in range(0, 60, 10) for _ in range(10)],
})
df.sample(5)
```

![Image](https://github.com/user-attachments/assets/7e15cf29-1d35-402b-ad87-34ea6c410e31)

**Initial problem** -> `cmap` is ignored
```python
df.hvplot.line('x', 'y', by='z', cmap='Reds')
```

![Image](https://github.com/user-attachments/assets/dfdf1445-badc-423b-a9d0-227297166cc9)

However, if you set `datashade=True` in the plot, the `cmap` keyword is consumed (somewhat)
```python
df.hvplot.line('x', 'y', by='z', cmap='Reds', datashade=True, dynspread=True, threshold=1)
```

![Image](https://github.com/user-attachments/assets/0919af10-7525-45a8-a21f-60d8192ea64e)

```python
df.hvplot.line('x', 'y', by='z', cmap='Set1', datashade=True, dynspread=True, threshold=1)
```
![Image](https://github.com/user-attachments/assets/0c12cd23-4bc3-4887-b93b-95e90805a349)

However, it doesn't make sense to use datashader for line plots just to get your preferred `cmap` used. So another way you can get hvplot to use the `cmap` is to plot with another element type - in this case, a points or scatter plot:

```python
df.hvplot.points('x', 'y', color='z', cmap='Reds')
```

![Image](https://github.com/user-attachments/assets/611c9ffc-230b-413e-a4f1-3fd7930deb23)

Notice however, that I used the `color` keyword instead of the `by` keyword, and also, you get a colorbar since you are coloring the plot using your numerical column. However, if I add a categorical column to the dataframe and use it to color the plot:
```python
df['cat'] = 'cat' + df['z'].astype(str)
df.hvplot.points('x', 'y', color='cat', cmap='Reds')
```

![Image](https://github.com/user-attachments/assets/8890336a-d51f-4853-a8a5-0db8bf531b8e)

Now you get a legend instead of a numerical colorbar. Doing the same with the original line plot will generate a ValueError:

```python
df.hvplot.line('x', 'y', color='z', cmap='Reds')
```
```
ValueError [Call holoviews.ipython.show_traceback() for details]
Mapping a dimension to the "color" style option is not supported by the Curve element using the bokeh backend. To map the "z" dimension to the color use a groupby operation to overlay your data along the dimension.
```

In conclusion, I think the colormap you get depends on the type of plot you are constructing (line vs scatter/points), the data type of the dimension you are using to color the plot (numerical or categorical), and/or the keyword you are using (`by` vs `color`).

**Comment by Azaya89** [comment]
The way I see it, there are a few changes that we can implement to make everything consistent.

1. We can document properly the difference between using `color` and `by` to color plots, which I believe has already been done in [this PR](#1574 )

![Image](https://github.com/user-attachments/assets/475b3dc6-6b6d-4a7d-8d7e-efccc7ca6c19)

2. We can add the feature as proposed by @maximlt [here](https://github.com/holoviz/hvplot/issues/652#issuecomment-2829853094) for only the `color` parameter since `color` uses vectorized colormapping instead of creating overlays like the `by` parameter.  This could be done somewhere in [converter.py](https://github.com/holoviz/hvplot/blob/main/hvplot/converter.py) and should work for both bokeh and matplotlib backends.
3. If the feature is added, we can make it available for as many plot elements as possible (line, scatter, points, etc)
4. We can add an error warning to places where a cmap is ignored (by + cmap) rather than just failing silently. Maybe something like this in converter.py
```python
if "by" in params and "cmap" in params:
         raise ValueError("`cmap` is only supported when using `color`, not `by`")
```

In summary, something like `df.hvplot.line('x', 'y', color='z', cmap='Reds')` should work but `df.hvplot.line('x', 'y', by='z', cmap='Reds')` should raise a clear error/warning.
